### PR TITLE
Align parser for swan lake object changes

### LINKED ISSUE #25312: Align parser with new parser rules for object type
**Description:**

**Steps to reproduce:**

**Affected Versions:**

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by lochana-chathura**
## Purpose
* Align parser with new parser rules for object type
* Add error recovery support for object constructor expression
* Support parsing class annotation declarations
* Fix bugs in  class definition error recovery

Fixes #25312, #25284, #25187, #25038

## Approach
N/A

## Samples
N/A

## Remarks
N/A

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [x] Added necessary documentation  
  - [x] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/SyntaxErrors.java [hunk 1]
```diff
@@ -240,6 +240,8 @@ private static DiagnosticCode getErrorCode(SyntaxKind expectedKind) {
                 return DiagnosticErrorCode.ERROR_MISSING_JOIN_KEYWORD;
             case OUTER_KEYWORD:
                 return DiagnosticErrorCode.ERROR_MISSING_OUTER_KEYWORD;
+            case CLASS_KEYWORD:
+                return DiagnosticErrorCode.ERROR_MISSING_CLASS_KEYWORD;
```
Comment (rdulmina) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/SyntaxErrors.java:244]: `getErrorCode(Syntaxkind expectedKind)` will be removed soon. There is a new method introduced `getErrorCode(ParserRuleContext currentCtx)` Please add all new contexts which u introduce to that method as u add in getExpectedToken in ErrorHandler.

Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/SyntaxErrors.java:244]: I see. class-change branch haven't synced with this change from master it seems. Will fix once synced. Thanks!

Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/SyntaxErrors.java:244]: Cool

**Comment by rdhananjaya** [comment]
Shall we limit commit message size to 50 char

Please refer https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/diagnostics/DiagnosticErrorCode.java [hunk 2]
```diff
@@ -239,7 +240,7 @@
     // Expression followed by the start keyword must be a func-call, a method-call or a remote-method-call
     ERROR_INVALID_EXPRESSION_IN_START_ACTION("BCE207", "error.invalid.expression.in.start.action"),
     // Cannot have the  same qualifier twice
-    ERROR_SAME_OBJECT_TYPE_QUALIFIER("BCE208", "error.same.object.type.qualifier"),
+    ERROR_SAME_CLASS_TYPE_QUALIFIER("BCE208", "error.same.class.type.qualifier"),
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/diagnostics/DiagnosticErrorCode.java:243]: Should we rename this to ERROR_DUPLICATE_CLASS_TYPE_QUALIFIER? it seems consistent with the rest.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/AbstractParserErrorHandler.java [hunk 3]
```diff
@@ -225,6 +225,15 @@ protected ParserRuleContext getGrandParentContext() {
         return grandParent;
     }
 
+    protected ParserRuleContext getGreatGrandParentContext() {
```
Comment (SupunS) [inline compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/AbstractParserErrorHandler.java:228]: Lets see if we can remove this

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/SyntaxErrors.java (hunk 1 above)
Comment (SupunS) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/SyntaxErrors.java:244]: @rdulmina Shall we remove that `getErrorCode(Syntaxkind expectedKind)` soon?

File: compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/declarations/ObjectTypeDefinitionTest.java [hunk 4]
```diff
@@ -79,72 +79,72 @@ public void testClientQualifierOnly() {
         test("object-type-def/object_type_def_source_14.bal", "object-type-def/object_type_def_assert_14.json");
     }
 
-    @Test
+    @Test(enabled = false) // no longer valid, due to spec deviation
```
Comment (SupunS) [inline compiler/ballerina-parser/src/test/java/io/ballerinalang/compiler/parser/test/syntax/declarations/ObjectTypeDefinitionTest.java:82]: Lets remove these invalid tests. May be we can keep one or two as negative test cases.

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/AbstractParserErrorHandler.java (hunk 3 above)
Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/AbstractParserErrorHandler.java:228]: yah. Haven't pushed the commit with code review suggestions yet.

**Review by rdhananjaya — APPROVED** [review]

File: compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/SyntaxErrors.java (hunk 1 above)
Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/SyntaxErrors.java:244]: @SupunS sure. Here is the issue created https://github.com/ballerina-platform/ballerina-lang/issues/25611

Comment (lochana-chathura) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/SyntaxErrors.java:244]: @rdulmina added to `getErrorCode(ParserRuleContext currentCtx)` with b3b7d5b in PR #25623

Comment (rdulmina) [reply compiler/ballerina-parser/src/main/java/io/ballerinalang/compiler/internal/parser/SyntaxErrors.java:244]: awesom

**Pull request merged**